sqlmesh/core/model/definition.py

```diff
@@ -1784,6 +1785,84 @@ class AuditResult(PydanticModel):
     blocking: bool = True
 
 
+def create_model_blueprints(
+    gateway: t.Optional[str | exp.Expression],
+    blueprints: t.Any,
+    get_variables: t.Callable[[t.Optional[str]], t.Dict[str, str]],
+    loader: t.Callable[..., Model],
+    **loader_kwargs: t.Any,
+) -> t.List[Model]:
+    path = loader_kwargs.get("path")
+    dialect = loader_kwargs.get("dialect")
+
+    if not blueprints:
+        blueprints = [None]
+    elif isinstance(blueprints, exp.Paren):
+        blueprints = [blueprints.unnest()]
+    elif isinstance(blueprints, (exp.Tuple, exp.Array)):
+        blueprints = blueprints.expressions
+    elif not isinstance(blueprints, list):
+        # TODO: put docs link here
+        raise ConfigError(
+            f"Invalid 'blueprints' property '{blueprints}' at '{path}', please refer to <link>."
+        )
+
+    model_blueprints: t.List[Model] = []
+    for blueprint in blueprints:
+        if not blueprint:
+            variables = {}
+        elif isinstance(blueprint, exp.Paren):
+            blueprint = blueprint.unnest()
+            variables = {blueprint.left.name: blueprint.right.sql(dialect=dialect)}
+        elif isinstance(blueprint, (exp.Tuple, exp.Array)):
+            variables = {e.left.name: e.right.sql(dialect=dialect) for e in blueprint.expressions}
+        elif isinstance(blueprint, dict):
+            variables = blueprint
+        else:
+            # TODO: put docs link here
+            raise ConfigError(
+                f"Invalid blueprint value: {blueprint} at '{path}', please refer to <link>."
+            )
+
+        if isinstance(gateway, d.MacroVar):
+            gateway_name = variables.get(gateway.name)
+        elif isinstance(gateway, str) and gateway.startswith("@"):
```

Comment:
why is gateway expected to be @? can someone pass in a normal gateway?

Reply:
They can, this is only supposed to handle the case where someone does `gateway @blueprint_value`, in order to fetch the variables of the corresponding gateway.

Reply:
I guess we want this logic to kick in anyway so that the appropriate variables are passed to the model?

Reply:
can gateway be a function call? @get_gateway()

Reply:
Yes, that is valid today. It's also valid in this PR, it simply won't be picked up in the `variables` dict lookup, see other [thread](https://github.com/TobikoData/sqlmesh/pull/3848#discussion_r1959510837).